README.md
# homework3_4

history.txt
Пн 15.02.2021 и время 11:45:21 PM MSK
Пн 15.02.2021 и время 11:50:45 PM MSK
Пн 15.02.2021 и время 11:54:53 PM MSK
Вт 16.02.2021 и время 12:11:24 AM MSK
Server: /clients vano 
vano: hello all
Вт 16.02.2021 и время 12:16:36 AM MSK
Server: /clients vano 
vano: hello
vano: how are you?
Вт 16.02.2021 и время 01:09:57 AM MSK
Вт 16.02.2021 и время 01:11:27 AM MSK
Server: /clients vano 
vano: привет. провряем connection timeout после авторизации
Вт 16.02.2021 и время 01:19:09 AM MSK
Server: /clients vano 
Вт 16.02.2021 и время 01:21:52 AM MSK
Вт 16.02.2021 и время 01:22:29 AM MSK
Server: /clients vano 
Вт 16.02.2021 и время 01:25:38 AM MSK
Server: /clients vano 
Вт 16.02.2021 и время 01:28:47 AM MSK
Server: /clients vano 
Вт 16.02.2021 и время 01:29:20 AM MSK
Вт 16.02.2021 и время 01:35:06 AM MSK
Вт 16.02.2021 и время 01:35:31 AM MSK
Вт 16.02.2021 и время 01:37:39 AM MSK
Вт 16.02.2021 и время 01:38:10 AM MSK
Вт 16.02.2021 и время 01:38:57 AM MSK
Сб 06.03.2021 и время 03:58:06 PM MSK


src/main/java/FirstTask/Lock.java
package FirstTask;

public class Lock {
    private volatile String task = "A";

    public synchronized void printA() {
        synchronized (task) {
            for (int i = 0; i < 5; i++) {
                while (!"A".equals(task)) {
                    try {
                        wait();
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
                System.out.print(task);
                task = "B";
                notifyAll();
            }
        }
    }

    public synchronized void printB() {
        synchronized (task) {
            for (int i = 0; i < 5; i++) {
                while (!"B".equals(task)) {
                    try {
                        wait();
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
                System.out.print(task);
                task = "C";
                notifyAll();
            }
        }
    }

    public synchronized void printC() {
        synchronized (task) {
            for (int i = 0; i < 5; i++) {
                while (!"C".equals(task)) {
                    try {
                        wait();
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
                System.out.print(task);
                task = "A";
                notifyAll();
            }
        }
    }
}


src/main/java/Server/DBAccess.java
// Структура БД
// БД - users
// таблица - users
// поля:
// ID - int, autoincrement, notnull, primary key
// LOGIN - varchar(20), notnull
// PASSWORD - varchar(20), notnull
// NICK - varchar(20), notnull

package Server;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.*;

public class DBAccess {
    private final String url = "jdbc:sqlite:src/main/resources/users.db";
    //private final String url = "jdbc:mysql://localhost:3306/users";
    private final String login = "mysqluser";
    private final String password = "paracels";
    private final String auth = "SELECT NICK FROM users WHERE LOGIN=? AND PASSWORD=?";
    private final String isNickBusy = "SELECT * FROM users WHERE NICK=?";
    private final String changeNick = "UPDATE users SET NICK=? WHERE NICK=?";
    private Connection connection = null;
    private String error = "";
    private static final Logger LOG = LogManager.getLogger(DBAccess.class);

    public String getError() {
        return error;
    }

    public void connect() {
        try {
            connection = DriverManager.getConnection(url); // for SQLite
            //connection = DriverManager.getConnection(url, login, password); // for MySQL
            LOG.info("Connection to SQLite has been established.");
            //System.out.println("Connection to MySql has been established.");
        }catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public String getNick(String login, String password) {
        try {
            String nick = "";
            PreparedStatement ps = connection.prepareStatement(auth);
            ps.setString(1, login);
            ps.setString(2, password);
            ResultSet rs = ps.executeQuery();
            while (rs.next()) {
                nick = rs.getString(1);
                System.out.println(nick);
            }
            if(!nick.isEmpty()) {
                return nick;
            }
        }catch (SQLException e) {
            e.printStackTrace();
        }
        return null;
    }

    public boolean changeNick(String oldNick, String newNick) {
        try {
            PreparedStatement ps = connection.prepareStatement(isNickBusy);
            ps.setString(1, newNick);
            ResultSet rs = ps.executeQuery();
            if(rs.next()) {
                this.error = "Ник занят";
                return false;
            }
            ps = connection.prepareStatement(changeNick);
            ps.setString(1, newNick);
            ps.setString(2, oldNick);
            ps.execute();
            return true;
        }catch (SQLException e) {
            e.printStackTrace();
        }
        return false;
    }

    public void closeConnection() {
        try {
            if(connection != null) {
                LOG.info("Connection to SQLite close.");
                connection.close();
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }
}


src/main/java/Server/MyServer.java
package Server;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MyServer {
    public static final int PORT = 8081;
    public static final int TIMEOUT = 120;
    private List<ClientHandler> clients;
    private AuthService authService;
    private ExecutorService executorService;
    private static final Logger LOG = LogManager.getLogger(MyServer.class);

    public MyServer() {
        try (ServerSocket serverSocket = new ServerSocket(PORT)) {
            authService = new BaseAuthService();
            authService.start();
            clients = new ArrayList<>();
            executorService = Executors.newCachedThreadPool();
            LOG.info("Сервер запущен");

            while (true) {
                System.out.println("Ожидаем подключение клиентов");
                Socket socket = serverSocket.accept();
                LOG.info("Клиент подключился");
                executorService.execute(new Thread(() -> new ClientHandler(this, socket)));
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            closeMyServer();
        }
    }

    public synchronized void sendMessageToClient(ClientHandler from, String nickTo, String msg) {
        for (ClientHandler client : clients) {
            if (client.getNick().equals(nickTo)) {
                String infoStr = "Отправляем личное сообщение от " + from.getNick() + ", кому " + nickTo;
                LOG.info(infoStr);
                Message message = new Message();
                message.setNick(from.getNick());
                message.setMessage(msg);
                client.sendMessage(message);
                return;
            }
        }
        String infoStr = "Клиент с ником " + nickTo +  "не подключен к чату";
        LOG.warn(infoStr);
        Message message = new Message();
        message.setMessage(infoStr);
        from.sendMessage(message);
    }

    public synchronized void broadcastMessage(Message message) {
        LOG.trace("Рассылка всем клиентам");
        for (ClientHandler client : clients) {
            client.sendMessage(message);
        }
    }

    public synchronized boolean isNickBusy(String nick) {
        for (ClientHandler client : clients) {
            if(nick.equals(client.getNick())) {
                return true;
            }
        }
        LOG.warn(nick + " занят");
        return false;
    }

    public AuthService getAuthService() {
        return authService;
    }

    public synchronized void subscribe(ClientHandler clientHandler) {
        clients.add(clientHandler);
        broadcastClientList();
        LOG.trace("клиент авторизовался");
    }

    public synchronized void unsubscribe(ClientHandler clientHandler) {
        clients.remove(clientHandler);
        broadcastClientList();
        LOG.trace("клиент вышел");
    }

    private synchronized void broadcastClientList() {
        StringBuilder sb = new StringBuilder("/clients ");
        for (ClientHandler client : clients) {
            sb.append(client.getNick()).append(" ");
        }
        Message message = new Message();
        message.setMessage(sb.toString());
        broadcastMessage(message);
        LOG.trace("рассылка списка всех подключенных клиентов");
    }

    private void closeMyServer() {
        if(authService != null) {
            authService.stop();
        }
        executorService.shutdown();
        LOG.info("завершение работы сервера");
    }
}

src/main/java/Server/ClientHandler.java
package Server;

import com.google.gson.Gson;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.Socket;
import java.util.concurrent.*;

public class ClientHandler {
    private Socket socket;
    private MyServer myServer;
    private DataInputStream dataInputStream;
    private DataOutputStream dataOutputStream;
    private String nick;
    private final ExecutorService executorService = Executors.newSingleThreadExecutor();
    private final Logger LOG = LogManager.getLogger(ClientHandler.class);

    @SuppressWarnings("rawtypes")
    public ClientHandler(MyServer myServer, Socket socket) {
        try {
            this.myServer = myServer;
            this.socket = socket;

            this.dataInputStream = new DataInputStream(socket.getInputStream());
            this.dataOutputStream = new DataOutputStream(socket.getOutputStream());

            Runnable auth = this::authentication;
            Future future = executorService.submit(auth);

            try {
                future.get(MyServer.TIMEOUT, TimeUnit.SECONDS);
                executorService.execute(this::readMessages);
            } catch (InterruptedException | ExecutionException e) {
                e.printStackTrace();
            } catch (TimeoutException e) {
                LOG.error("Connection timeout "+MyServer.TIMEOUT+" sec");
                closeConnection();
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void closeConnection() {
        Message message = new Message();
        myServer.unsubscribe(this);
        message.setMessage(nick + " вышел из чата");
        myServer.broadcastMessage(message);
        try {
            dataOutputStream.close();
            dataInputStream.close();
            socket.close();
            LOG.info(nick + ": сеанс завершен");
        } catch (IOException ignore) {
        }
        executorService.shutdown();
    }

    private void authentication() {
        while (true) {
            try {
                AuthMessage message = new Gson().fromJson(dataInputStream.readUTF(), AuthMessage.class);
                String nick = myServer.getAuthService().getNickByLoginAndPass(message.getLogin(), message.getPassword());
                if (nick != null && !myServer.isNickBusy(nick)) {
                    message.setAuthenticated(true);
                    message.setNick(nick);
                    this.nick = nick;

                    dataOutputStream.writeUTF(new Gson().toJson(message));
                    Message broadcastMsg = new Message();
                    broadcastMsg.setMessage(nick + " вошел в чат");
                    myServer.broadcastMessage(broadcastMsg);
                    myServer.subscribe(this);
                    return;
                } else {
                    message.setAuthenticated(false);
                    message.setNick("/wrong");
                    dataOutputStream.writeUTF(new Gson().toJson(message));
                }
            } catch (IOException ignored) {
            }
        }
    }

    private void readMessages() {
        while (true) {
            Message message = null;
            try {
                message = new Gson().fromJson(dataInputStream.readUTF(), Message.class);
            } catch (IOException e) {
                closeConnection();
            }
            if(message != null) {
                message.setNick(nick);

                System.out.println(message);
                if (!message.getMessage().startsWith("/")) {
                    myServer.broadcastMessage(message);
                    continue;
                }
                String[] command = message.getMessage().split("\\s");
                switch (command[0]) {
                    case "/end": {
                        LOG.info("Close connection from " + nick);
                        closeConnection();
                        return;
                    }
                    case "/w": {
                        if (command.length < 3) {
                            Message msg = new Message();
                            msg.setMessage("Не хватает параметров, необходимо отправить команду следующего вида: /w <ник> <сообщение>");
                            LOG.warn("Не хватает параметров, необходимо отправить команду следующего вида: /w <ник> <сообщение>");
                            this.sendMessage(msg);
                        }
                        String nick = command[1];
                        StringBuilder sbMessage = new StringBuilder();
                        for (int i = 2; i < command.length; i++) {
                            sbMessage.append(command[i]);
                            sbMessage.append(" ");
                        }
                        myServer.sendMessageToClient(this, nick, sbMessage.toString());
                        break;
                    }
                    case "/ch": {
                        if (command.length != 2) {
                            Message msg = new Message();
                            msg.setMessage("Необходимо отправить команду следующего вида:\n/ch <новый_ник>");
                            LOG.warn("Необходимо отправить команду следующего вида:\n/ch <новый_ник>");
                            this.sendMessage(msg);
                            break;
                        }
                        if (!myServer.getAuthService().changeNick(nick, command[1])) {
                            Message msg = new Message();
                            msg.setMessage(myServer.getAuthService().getError());
                            this.sendMessage(msg);
                            break;
                        }
                        Message msg = new Message();
                        msg.setMessage(this.nick + " теперь известен как " + command[1]);
                        LOG.info(this.nick + " теперь известен как " + command[1]);
                        myServer.broadcastMessage(msg);
                        this.nick = command[1];
                        break;
                    }
                    case "/h": {
                        Message msg = new Message();
                        String sb = "\n/w <nick> <message> \t Отправить личное сообщение\n" +
                                "/ch <new_nick> \t Сменить ник (если свободен)\n" +
                                "/h \t\t Это сообщение\n" +
                                "/end \t\t Завершить сеанс\n";
                        msg.setMessage(sb);
                        this.sendMessage(msg);
                        break;
                    }
                }
            }
        }
    }

    public void sendMessage(Message message) {
        try {
            dataOutputStream.writeUTF(new Gson().toJson(message));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public String getNick() {
        return nick;
    }
}


src/main/java/Server/BaseAuthService.java
package Server;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class BaseAuthService implements AuthService {

    private final DBAccess dbAccess;
    private static final Logger LOG = LogManager.getLogger(BaseAuthService.class);

    public BaseAuthService() {
        dbAccess = new DBAccess();
    }

    @Override
    public void start() {
        dbAccess.connect();
        LOG.info("Сервис авторизации запущен");
    }

    @Override
    public void stop() {
        dbAccess.closeConnection();
        LOG.info("Сервис авторизации остановлен");
    }

    @Override
    public String getNickByLoginAndPass(String login, String password) {
        return dbAccess.getNick(login, password);
    }

    @Override
    public boolean changeNick(String oldNick, String newNick) {
        return dbAccess.changeNick(oldNick, newNick);
    }

    @Override
    public String getError() {
        return dbAccess.getError();
    }
}


src/main/java/Client/MyClient.java
package Client;

import Server.Message;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.IOException;

public class MyClient extends JFrame {

    private final ServerService serverService;
    private final LocalHistory localHistory;

    public MyClient() {
        super("Simple Chat");

        localHistory = new LocalHistory();
        serverService = new SocketServerService();
        serverService.openConnection();

        setDefaultCloseOperation(WindowConstants.DO_NOTHING_ON_CLOSE);
        this.addWindowListener(new WindowListener() {
            public void windowActivated(WindowEvent event) {
            }
            public void windowClosed(WindowEvent event) {
            }
            public void windowClosing(WindowEvent event) {
                exitApp(event);
            }
            public void windowDeactivated(WindowEvent event) {
            }
            public void windowDeiconified(WindowEvent event) {
            }
            public void windowIconified(WindowEvent event) {
            }
            public void windowOpened(WindowEvent event) {
            }
        });

        setLayout(new FlowLayout());
        setBounds(400, 400, 520, 350);
        setResizable(false);


        JTextArea mainChat = new JTextArea();
        mainChat.setSize(100, 400);
        mainChat.setColumns(45);
        mainChat.setRows(15);
        mainChat.setFocusable(false);
        mainChat.setEditable(false);
        mainChat.setLineWrap(true);
        mainChat.setBorder(BorderFactory.createEtchedBorder());
        JScrollPane scroll = new JScrollPane (mainChat,
                JScrollPane.VERTICAL_SCROLLBAR_AS_NEEDED, JScrollPane.HORIZONTAL_SCROLLBAR_NEVER);

        JTextField myMessage = new JTextField();
        myMessage.setSize(100, 400);
        myMessage.setColumns(45);
        myMessage.setBorder(BorderFactory.createEtchedBorder());

        Label loginLabel = new Label("Login");
        JTextField login = new JTextField();
        login.setColumns(10);
        Label passwordLabel = new Label("Password");
        JPasswordField password = new JPasswordField();
        password.setColumns(10);

        JLabel status = new JLabel("Status: Не авторизован");

        JButton send = new JButton("Send");
        send.setSize(50, 200);

        send.addActionListener(actionEvent -> sendAuth(loginLabel, login, passwordLabel, password,
                send, mainChat, myMessage, status));

        myMessage.addKeyListener(new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent e) {
                if(e.getKeyCode() == KeyEvent.VK_ENTER) {
                    sendMessage(myMessage);
                }
            }
        });

        JButton exitButton = new JButton("Exit");
        exitButton.setSize(50, 200);
        exitButton.addActionListener(actionEvent -> exitButton());

        add(status);

        //add(mainChat);
        add(scroll);
        add(myMessage);

        add(loginLabel);
        add(login);
        add(passwordLabel);
        add(password);

        add(send);
        add(exitButton);

        myMessage.setEditable(false);
        if(!serverService.getError().isEmpty()) {
            mainChat.append(serverService.getError());
        }
    }

    private void exitApp(WindowEvent event) {
        Object[] options = { "Да", "Нет!" };
        int n = JOptionPane
                .showOptionDialog(event.getWindow(), "Закрыть окно?",
                        "Подтверждение", JOptionPane.YES_NO_OPTION,
                        JOptionPane.QUESTION_MESSAGE, null, options,
                        options[0]);
        if (n == 0) {
            serverService.sendMessage("/end");
            localHistory.close();
            serverService.closeConnection();
            System.exit(0);
        }
    }

    private void sendAuth(Label loginLabel, JTextField login, Label passwordLabel, JPasswordField password,
                          JButton send, JTextArea mainChat, JTextField myMessage, JLabel status) {
        System.out.println("Попытка авторизации");

        if(login.getText().isEmpty() || String.valueOf(password.getPassword()).isEmpty()) {
            mainChat.append("System: Необходимо авторизоваться!\n");
            return;
        }

        try {
            serverService.authentication(login.getText(), String.valueOf(password.getPassword()));
        } catch (IOException e) {
            e.printStackTrace();
        }
        if(serverService.isConnected()) {
            loginLabel.setVisible(false);
            login.setVisible(false);
            passwordLabel.setVisible(false);
            password.setVisible(false);
            myMessage.setEditable(true);
            status.setText("Status: Авторизован");
            for(ActionListener listener : send.getActionListeners()) {
                send.removeActionListener(listener);
            }

            send.addActionListener(actionEvent -> sendMessage(myMessage));
            mainChat.append(localHistory.readHistory(100));

            new Thread(() -> {
                while (true) {
                    printToUI(mainChat, serverService.readMessages());
                }
            }).start();
        } else if (!serverService.getError().isEmpty()) {
            mainChat.append(serverService.getError());
        } else {
            mainChat.append("System: Авторизация не прошла\n");
        }
    }

    private void exitButton() {
        serverService.sendMessage("/end");
        localHistory.close();
        serverService.closeConnection();
        System.exit(0);
    }

    private void sendMessage(JTextField myMessage) {
        if(!myMessage.getText().isEmpty()) {
            serverService.sendMessage(myMessage.getText());
            myMessage.setText("");
        }
    }

    private void printToUI(JTextArea mainChat, Message message) {
        if(message.getMessage() != null) {
            String msg = ((message.getNick() == null) ? "Server" : message.getNick()) + ": " + message.getMessage();
            mainChat.append(msg + '\n');
            localHistory.writeHistory(msg);
        } else {
            mainChat.append("System: Что-то пошло не так\n");
        }
    }
}
